pkg/systemd/logs.js

```diff
@@ -192,7 +199,59 @@ $(function() {
             }
         }
 
+        function clear_service_list() {
+            $('#journal-services-list').empty()
+                    .append($('<li>').append($('<a>')
+                            .text(_("All"))
+                            .attr('data-service', "")))
+                    .append($('<li>')
+                            .addClass("divider"));
+        }
+
+        function load_service_filters(match, options, stopped) {
+            var tags_count = 0;
+            if (stopped) {
+                options["reverse"] = true;
+                options["cursor"] = stopped;
+            }
+            procs.push(journal.journalctl(match, options)
```

Comment:
This function iterates over the entire journal (which is really slow), while later in the outer journalbox() function it does it again. This is too expensive, can this please be merged into one loop? It's fine to render the journal list first and then update the service filter.

Reply:
> which is really slow

Not the fastest thing, but it does not seem to be that expensive as I was afraid it could be.

> while later in the outer journalbox() function it does it again

You mean this 
https://github.com/cockpit-project/cockpit/blob/master/pkg/systemd/logs.js#L219 ?
The problem with this is that if once selected a service then some services from first N logs may be lost. That if I would use first 5000  with this and then continue reading logs on its own, then all services that only happened in the first 5000 are not shown.

Reply:
Ah, because these two have different `match` arguments, I see. So I agree that these two need to be separate.

But still, instead of reading thousands of logs and parsing them on the JavaScript side, couldn't this do the API  equivalent of `journalctl --field=SYSLOG_IDENTIFIER`? Then this will filter on the server side, which should be much faster and simpler? This can be a follow-up, too, but the logs page is already slow enough on load as it is..

Reply:
so, normally the command we use looks like this `journalctl -q --output=json --no-tail --since=-1days --reverse -- PRIORITY=0 PRIORITY=1 PRIORITY=2 PRIORITY=3`
when I added `--field=SYSLOG_IDENTIFIER` it is not supported with matching.


    # journalctl -q --output=json --no-tail --since=-1days --reverse --field=SYSLOG_IDENTIFIER -- PRIORITY=0 PRIORITY=1 PRIORITY=2 PRIORITY=3

    Extraneous arguments starting with 'PRIORITY=0'

Surely I can use `-p 3` instead of this matching (which btw. maybe I should do in follow up), but then all possible values of syslog identifier in all logs are returned, ignoring other switches such as priority or time restriction. This list would contain mostly services, which when selected would return empty list.

Another thing that can be done is using `--output-field=SYSLOG_IDENTIFIER` to limit amount of data that is being transferred.  This should be fairly simple. I will take a look at this.

The last thing that comes to my mind is to use command with `output-field` and through pipes create list of services (pick just syslog_identifier as there is for example also boot_id always present, then make it unique list...) but I am not sure how this would work, but I can try to make it work.